lib/urlwatch/jobs.py

```diff
@@ -259,6 +268,12 @@ class UrlJob(Job):
     def get_location(self):
         return self.user_visible_url or self.url
 
+    def set_location(self, location):
+        if self.user_visible_url:
+            self.user_visible_url = location
+        else:
+            self.url = location
```

Comment:
This means that if a job has a `user_visible_url` set, the "real" URL isn't updated? Not sure if that's what a user would expect. Seems like user visible URL is something that's only for "cosmetics" (display), and the `--change-location` feature should always work on the "real" / "watched" URL? (same comment applies to the other `set_location()` functions).

Reply:
Yes, this threw me for a loop a bit.  I expected `user_visible_url` to simply be cosmetic, but it seems as though if `user_visible_url` is set, _that_ gets [used](https://github.com/thp/urlwatch/blob/c53876b02597d76e59ca2d0a15a7eb5ef0573fca/lib/urlwatch/jobs.py#L260) to compute the GUID, not the "true" URL.  I'm not too sure why the `user_visible_url` is used as the `location` to compute the GUID, but since it is, it seemed like that should be the value modified when editing the `location`

Hmmm, though now that I think about it more, if we leave the `user_visible_url` untouched and only ever modify the true URL then the job history _will_ be retained since the GUID will also be untouched in that situation.  The only downside is that for jobs with `user_visible_url` set, `get_location()` will not return the string passed in with a previous call to `set_location()`.  I suppose `set_location()` could be renamed though I'm not sure of a name that would make sense across all three job types.

Reply:
The thing with `user_visible_url` is that it was added as a "cosmetic" thing. Probably it accidentally got used as base for the GUID.

Here's one idea:

- Introduce an `original_guid` field that's empty by default and gets set by this migration
- If `user_visible_url` is set, but `original_guid` is not, it's used as the basis (like now, for backwards compatibility)
- If `original_guid` is set, it always "wins" over any other setting

This way, the migration only needs to set `original_guid` and it will work properly, and there's not even the need to rewrite the history, and existing behaviour isn't changed as long as original_guid is not set.

Reply:
Yes, that would work though I'm not a big fan of writing the GUID in the config file.  I know I'd be surprised to see a long random string show up in my config.  I have also benefited in the past from being able to manually compute a GUID from a url string in tests and one-off scripts.  Storing an original GUID based on some previous value of the url string would make it harder to do that conversion manually since it sort of "overrides" the GUID computation.

I think this achieves the same result without "saving" the GUID:

- Rename `set_location()` to `set_base_location()` and have it only ever set `self.url`, `self.command`, or `self.navigate` according to the job type. `user_visible_url` is untouched
- `get_location()` remains unchanged for backwards compatibility

That should result in the following behavior:

- If a job has `user_visible_url` set and this migration is executed, the `(url|command|navigate)` value is updated as one would expect.  The GUID never changes since `user_visible_url` is still the basis. History is retained.
- If a job does _not_ have `user_visible_url` set and this migration is executed, the `(url|command|navigate)` value is updated as one would expect and the GUID changes too.  The history is edited to match and is therefore retained.

What do you think?

Reply:
That sounds sensible, let's go with that. Also it should be backwards compatible and not be a behavior change compared to what was before (people adding a `user_visible_url` after the fact would still get their history lost, except if it's used as a "watchable URL has changed, but we want to keep history, so we'll put the old user visible URL there", for which this feature now exists).

Reply:
Cool, I just implemented that.  Thanks for the thoughtful discussion.